iceaxe/alias_values.py

```
@@ -1,36 +1,34 @@
+from dataclasses import dataclass
 from typing import Generic, TypeVar, cast
 
+from iceaxe.functions import FunctionMetadata
+
 T = TypeVar("T")
 
 
+@dataclass(frozen=True, slots=True)
 class Alias(Generic[T]):
-    def __init__(self, name: str, type: T):
-        self.name = name
-        self.type = type
+    name: str
+    value: T | FunctionMetadata
 
     def __str__(self):
         return self.name
 
 
-def alias(name: str, type: T) -> T:
+def alias(name: str, type: T | FunctionMetadata) -> T:
```

Since T is a generic TypeVar, it will supersede the value of FunctionMetadata here. What are you intending it to model as part of the typehint signature?

---

As far as I can tell, the original alias you implemented extracted values from existing `AS` statements so they could be returned by a select with the correct typing. My goal with this was to support that functionality, along with the ability to effectively write these `AS` statements inline with correct typing.

For the moment, I will remove `FunctionMetadata` from the typing, but I might be beneficial to eventually switch to a `FunctionMetadata` only approach in the future. If someone wants to use raw sql, they can just do it inline.

---

I totally forgot that this is the error which prompted me to put it in in the first place.
```
/Users/matthewlemay/Github/iceaxe/iceaxe/__tests__/test_text_search.py
  /Users/matthewlemay/Github/iceaxe/iceaxe/__tests__/test_text_search.py:174:9 - error: No overloads for "select" match the provided arguments (reportCallIssue)
  /Users/matthewlemay/Github/iceaxe/iceaxe/__tests__/test_text_search.py:174:26 - error: Argument of type "tuple[type[Article], FunctionMetadata]" cannot be assigned to parameter "fields" of type "tuple[T@select | type[T@select], T2@select | type[T2@select], T3@select | type[T3@select], T4@select | type[T4@select], T5@select | type[T5@select], T6@select | type[T6@select], T7@select | type[T7@select], T8@select | type[T8@select], T9@select | type[T9@select], T10@select | type[T10@select], *Ts@select]" in function "select"
    Type "FunctionMetadata" is not assignable to type "T2@select | type[T2@select]"
      Type "FunctionMetadata" is not assignable to type "SUPPORTED_SELECTS"
        Type "FunctionMetadata" is not assignable to type "SUPPORTED_SELECTS"
          "FunctionMetadata" is not assignable to "TableBase"
          "FunctionMetadata" is not assignable to "DBModelMetaclass"
          Type "FunctionMetadata" is not assignable to type "type[Enum]"
          Type "FunctionMetadata" is not assignable to type "type[StrEnum]"
          Type "FunctionMetadata" is not assignable to type "type[IntEnum]" (reportArgumentType)
  /Users/matthewlemay/Github/iceaxe/iceaxe/__tests__/test_text_search.py:278:9 - error: No overloads for "select" match the provided arguments (reportCallIssue)
  /Users/matthewlemay/Github/iceaxe/iceaxe/__tests__/test_text_search.py:278:26 - error: Argument of type "tuple[type[Article], FunctionMetadata]" cannot be assigned to parameter "fields" of type "tuple[T@select | type[T@select], T2@select | type[T2@select], T3@select | type[T3@select], T4@select | type[T4@select], T5@select | type[T5@select], T6@select | type[T6@select], T7@select | type[T7@select], T8@select | type[T8@select], T9@select | type[T9@select], T10@select | type[T10@select], *Ts@select]" in function "select"
    Type "FunctionMetadata" is not assignable to type "T2@select | type[T2@select]"
      Type "FunctionMetadata" is not assignable to type "SUPPORTED_SELECTS"
        Type "FunctionMetadata" is not assignable to type "SUPPORTED_SELECTS"
          "FunctionMetadata" is not assignable to "TableBase"
          "FunctionMetadata" is not assignable to "DBModelMetaclass"
          Type "FunctionMetadata" is not assignable to type "type[Enum]"
          Type "FunctionMetadata" is not assignable to type "type[StrEnum]"
          Type "FunctionMetadata" is not assignable to type "type[IntEnum]" (reportArgumentType)
4 errors, 0 warnings, 0 informations
make: *** [lint-iceaxe] Error 1
```

---

Ah I see. I'll checkout this branch today and do some thinking of what typing / `@overload` would work best here.

---

@mplemay A bit more context here: this error type cascade is caused by our typehinted definitions for the allowable T parameters that can be used in selects:

```python
SUPPORTED_SELECTS = (
    TableBase
    | DBModelMetaclass
    | ALL_ENUM_TYPES
    | PRIMITIVE_TYPES
    | PRIMITIVE_WRAPPER_TYPES
    | DATE_TYPES
    | JSON_WRAPPER_FALLBACK
    | None
)

T = TypeVar("T", bound=SUPPORTED_SELECTS)
```

Since FunctionMetadata isn't actually returned by the database, it doesn't make much sense to include there. The `T | FunctionMetadata` workaround seemingly silenced this error by masking out the FunctionMetadata itself.

I've fixed by casting ts_rank's return value to an integer, which seems to be the only possible return value. Let me know if you see any strange edge cases with that - but assuming not, I'll give this one last look and then merge in.